### PR TITLE
add post

### PR DESCRIPTION
**Pull request by maelle**

### CODE REVIEW

**Comment by maelle** [comment]
Direct link to post preview https://deploy-preview-244--ropensci.netlify.com/blog/2018/08/28/birds-ocr/

**Comment by maelle** [comment]
I don't understand the absence of line breaks before some titles. 😿

**Comment by jeroen** [comment]
```
> spelling::spell_check_files("2018-08-28-birds-ocr.md")
  WORD         FOUND IN
australia    2018-08-28-birds-ocr.md:367
BHL          2018-08-28-birds-ocr.md:336,348
cld          2018-08-28-birds-ocr.md:12,13
flickr       2018-08-28-birds-ocr.md:34
Flickr       2018-08-28-birds-ocr.md:55
github       2018-08-28-birds-ocr.md:15
Jeroen       2018-08-28-birds-ocr.md:39
Maëlle      2018-08-28-birds-ocr.md:5
magick       2018-08-28-birds-ocr.md:10
md           2018-08-28-birds-ocr.md:15
ocr          2018-08-28-birds-ocr.md:2
Ooms’      2018-08-28-birds-ocr.md:39
OSD          2018-08-28-birds-ocr.md:132
paclage      2018-08-28-birds-ocr.md:350
Radolfzell   2018-08-28-birds-ocr.md:8
rOpenSci     2018-08-28-birds-ocr.md:315
rOpenSci’s 2018-08-28-birds-ocr.md:22,39,231,348
taxize       2018-08-28-birds-ocr.md:14
tesseract    2018-08-28-birds-ocr.md:11
Tesseract    2018-08-28-birds-ocr.md:122,128
th           2018-08-28-birds-ocr.md:8
toolbelt     2018-08-28-birds-ocr.md:211
wll          2018-08-28-birds-ocr.md:5
worfklow     2018-08-28-birds-ocr.md:342
yaml         2018-08-28-birds-ocr.md:15
```

**Comment by jeroen** [comment]
> I don't understand the absence of line breaks before some titles. 😿
Hmmm I'll look into that as well.

**Comment by maelle** [comment]
Thanks, by the way could your spelling code become a part of all PRs?

**Comment by jeroen** [comment]
I usually avoid using two headers right after one another. This is a bit of a personal preference, but I think you can omit the second header ("image preperation") here:

screen shot 2018-08-08 at 4 13 59 pm

**Comment by maelle** [comment]
@jeroen Thanks, I removed the typos, and added text between headers.

I have tried using one line break instead of two but it didn't work, some headers still appear on the line before... I can't see any difference between them and other, correctly-placed, headers! 🙀

File: content/blog/2018-08-28-birds-ocr.md [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+
+A long time ago I had bookmarked the [Flickr account of the Biodiversity
+Heritage Library](https://www.flickr.com/people/biodivlibrary/). So many
+beautiful images of biodiversity, moreover free to use! In particular, I
+[downloaded all pictures from one of the Birds of Australia
+albums](https://www.flickr.com/photos/biodivlibrary/sets/72157694191194992).
+
+I wanted to try to extract the bird names on images using packages of
+Jeroen Ooms’, rOpenSci’s post-doc hacker & C(++)-bindings wizard. In
+particular, I wanted to try extract names of
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: s/try extract names of magick/try to extract names using magick/ ?

File: content/blog/2018-08-28-birds-ocr.md [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+
+### Free natural history images and appropriate R tooling!
+
+A long time ago I had bookmarked the [Flickr account of the Biodiversity
+Heritage Library](https://www.flickr.com/people/biodivlibrary/). So many
+beautiful images of biodiversity, moreover free to use! In particular, I
+[downloaded all pictures from one of the Birds of Australia
+albums](https://www.flickr.com/photos/biodivlibrary/sets/72157694191194992).
+
+I wanted to try to extract the bird names on images using packages of
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: s/names on images/names from images/ ?

File: content/blog/2018-08-28-birds-ocr.md [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+particular, I wanted to try extract names of
+[`magick`](https://github.com/ropensci/magick) for image manipulation,
+[`tesseract`](https://github.com/ropensci/tesseract) for optical
+character recognition,
+[`cld2`](https://github.com/ropensci/cld2)/[`cld3`](https://github.com/ropensci/cld3)
+for language detection… Quite the armory! We’ll also sprinkle some
+taxonomy magic by Scott Chamberlain, one of rOpenSci’s co-founders, to
+resolve the names extracted.
+
+### OCR-bird naming workflow, piece by piece
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: omit hyphen
looks like it's "OCR-bird" though I know that's not what you mean

File: content/blog/2018-08-28-birds-ocr.md [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+    preserve_yaml: true
+---
+
+In this new post, we’re taking a break from modern birding data in [our
+birder’s series](https://ropensci.org/tags/birder)… let’s explore
+gorgeous drawings from a natural history collection! Armed with
+rOpenSci’s packages binding powerful C++ libraries and open taxonomy
+data, how much information can we automatically extract from images?
+Maybe not much, but we’ll at least have explored image manipulation,
+optical character recognition, language detection, taxonomic name
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: add (OCR) since abbrev is used in heading below without definition
search doc and use OCR where appropriate

File: content/blog/2018-08-28-birds-ocr.md [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+[`tesseract`](https://github.com/ropensci/tesseract) for optical
+character recognition,
+[`cld2`](https://github.com/ropensci/cld2)/[`cld3`](https://github.com/ropensci/cld3)
+for language detection… Quite the armory! We’ll also sprinkle some
+taxonomy magic by Scott Chamberlain, one of rOpenSci’s co-founders, to
+resolve the names extracted.
+
+### OCR-bird naming workflow, piece by piece
+
+In this section, we explain the different elements of our R workflow:
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: mixing of "I" and "we" confusing

File: content/blog/2018-08-28-birds-ocr.md [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+wrapper](https://github.com/jimhester/flickrr)… I don’t do *everything*
+with R scripts (yet).
+
+``` r
+library("magrittr")
+filenames <- fs::dir_ls("birds")
+```
+
+Each image shows a bird and its name. Images are either landscape- or
+portrait- oriented, but for the sake of simplicity, I’ll act as if they
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: s/portrait- oriented/portrait-oriented/
in 2 places

File: content/blog/2018-08-28-birds-ocr.md [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+were all portrait- oriented. *What I’d do if I weren’t after simplicity
+but still as lazy would be to duplicate images rotated
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: revised for reader clarity:
*What I’d do if I weren’t after simplicity
but still as lazy, would be to duplicate images rotated
(`magick::image_rotate`) in all possible directions and apply the
workflow, hoping to get good data from one of the rotated versions.*

File: content/blog/2018-08-28-birds-ocr.md [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+  return(image)
+}
+
+crop_bird(filenames[1])
+```
+
+<img src="/img/blog-images/2018-08-28-birds-ocr/unnamed-chunk-3-1.png" width="1376" />
+
+It does look cleaner now!
+
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: need line break before heading (I know these are trouble)

File: content/blog/2018-08-28-birds-ocr.md [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+and this [Wiki page of Tesseract about improving the quality of the
+output](https://github.com/tesseract-ocr/tesseract/wiki/ImproveQuality).
+The [`hocr` package](https://github.com/dmi3kno/hocr) might be of
+interest for post-processing of OCR results.
+
+Below, we only changed the [page segmentation
+mode](https://github.com/tesseract-ocr/tesseract/wiki/ImproveQuality#page-segmentation-method)
+choosing 1 for “Automatic page segmentation with Orientation and script
+detection (OSD)”. We tried using Latin instead of English training data,
+but the results were less good.
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: s/less/not as/

File: content/blog/2018-08-28-birds-ocr.md [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+
+One can use either `tesseract::ocr` for a file path, url, or raw vector
+to image, or `magick::image_ocr` for a `magick` object which is quite
+handy in our pipeline.
+
+The function below also filter results of the OCR using language
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: s/filter/filters/

File: content/blog/2018-08-28-birds-ocr.md [hunk 11]
```diff
@@ -0,0 +1,411 @@
...
+detection. By only keeping text recognized as either Latin or English by
+one of the two languages detection packages `cld2` and `cld3` that are
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: s/languages detection/language detection/

File: content/blog/2018-08-28-birds-ocr.md [hunk 12]
```diff
@@ -0,0 +1,411 @@
...
+  get_names())
+```
+
+    ## # A tibble: 2 x 4
+    ##   text                 cld2  cld3  filename                          
+    ##   <chr>                <chr> <chr> <fs::path>                        
+    ## 1 climacteris picumnus <NA>  la    birds/n115_w1150_42399797481_o.jpg
+    ## 2 brown tree creeper   en    <NA>  birds/n115_w1150_42399797481_o.jpg
+
+So we see that the image transformation was quite useful.
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: Sentence "So we see that the image transformation was quite useful." is a bit unclear after a long bunch of code snippets i.e. where does a novice coder look? Could you move that line above the tibble? or add a phrase giving example of "climacteris picumnus"?

File: content/blog/2018-08-28-birds-ocr.md [hunk 13]
```diff
@@ -0,0 +1,411 @@
...
+
+Now, these names look fine, but how to be sure they’re actually
+taxonomic names?
+
+#### Taxonomic name resolution
+
+The [`taxize` package](https://github.com/ropensci/taxize/) by Scott
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: with Zachary Foster now having maintainer status too, should he be mentioned even though Scott is major committer?

File: content/blog/2018-08-28-birds-ocr.md [hunk 14]
```diff
@@ -0,0 +1,411 @@
...
+Chamberlain, taxonomic toolbelt for R, provides access to many fantastic
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: maybe rephrase so Scott doesn't interrupt the package name and it's purpose:

"The [`taxize` package](https://github.com/ropensci/taxize/), a taxonomic toolbelt for R, by Scott Chamberlain..."

or "The [`taxize` package(https://github.com/ropensci/taxize/) by Scott Chamberlain, is a taxonomic toolbelt for R, providing access to many fantastic data sources and tools for taxonomy"

File: content/blog/2018-08-28-birds-ocr.md [hunk 15]
```diff
@@ -0,0 +1,411 @@
...
+data sources and tools for taxonomy. One of them, the Global Name
+Resolver, provides, well, resolution of taxonomic names, sadly not
+common names. `taxize::gnr_resolve` has many options, we only make use
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: s/options, we /options, but we /

File: content/blog/2018-08-28-birds-ocr.md [hunk 16]
```diff
@@ -0,0 +1,411 @@
...
+  best_match_only = TRUE)
+```
+
+    ## # A tibble: 1 x 5
+    ##   user_supplied_name submitted_name  matched_name   data_source_tit~ score
+    ## * <chr>              <chr>           <chr>          <chr>            <dbl>
+    ## 1 climacteris picum~ Climacteris pi~ Climacteris p~ NCBI             0.988
+
+English names could be cleaned up a bit using the [`spelling`
+package](https://github.com/ropensci/spelling), also an rOpenSci’s
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: s/an rOpenSci’s package/an rOpenSci package/

File: content/blog/2018-08-28-birds-ocr.md (hunk 3 above)
Comment (stefaniebutland) [reply content/blog/2018-08-28-birds-ocr.md]: same for heading below

File: content/blog/2018-08-28-birds-ocr.md [hunk 17]
```diff
@@ -0,0 +1,411 @@
...
+bird_names <- purrr::map(filenames, crop_bird) %>%
+  purrr::map_df(get_names)
+```
+
+Out of 51 images only 17 are present in the final table with possible
+names which is a bit disheartening, but we could surely do better in the
+image processing and OCR tweaking steps! Maybe one could frame the
+parameter search [as a machine learning
+problem](https://twitter.com/dmi3k/status/1024919690768990209). We
+should also keep in mind that such Natural History images are quite hard
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: no caps in Natural History?

File: content/blog/2018-08-28-birds-ocr.md [hunk 18]
```diff
@@ -0,0 +1,411 @@
...
+  dplyr::mutate(gnr = ifelse(cld2 == "la" | cld3 == "la",
+                             safe_resolve(text),
+                             list(NULL)))
+```
+
+We do not get much resolution, but we knew the names weren’t very good
+to start with. A better idea here might be to get a full list of names
+of Australian birds, potentially leveraging [the `taxizedb` package by
+Scott Chamberlain](https://github.com/ropensci/taxizedb), and to then
+fuzzy-match them with the names we have.
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: you should comment on whether that was helpful or not

File: content/blog/2018-08-28-birds-ocr.md [hunk 19]
```diff
@@ -0,0 +1,411 @@
...
+
+-   [`cld2`](https://github.com/ropensci/cld2)/[`cld3`](https://github.com/ropensci/cld3)
+    for language detection.
+
+We also used [a function from
+`taxize`](https://github.com/ropensci/taxize) allowing to use [the
+Global Name Resolver](https://resolver.globalnames.org/about). Discover
+more packages from our suite [here](https://ropensci.org/packages/).
+
+#### Applicability of this OCR bird-naming workflow
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: different hyphen use here. please resolve

File: content/blog/2018-08-28-birds-ocr.md [hunk 20]
```diff
@@ -0,0 +1,411 @@
...
+    manipulation,
+
+-   [`tesseract`](https://github.com/ropensci/tesseract) for optical
+    character recognition,
+
+-   [`cld2`](https://github.com/ropensci/cld2)/[`cld3`](https://github.com/ropensci/cld3)
+    for language detection.
+
+We also used [a function from
+`taxize`](https://github.com/ropensci/taxize) allowing to use [the
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: s/allowing to use/allowing us to use/ ?

File: content/blog/2018-08-28-birds-ocr.md [hunk 21]
```diff
@@ -0,0 +1,411 @@
...
+Global Name Resolver](https://resolver.globalnames.org/about). Discover
+more packages from our suite [here](https://ropensci.org/packages/).
+
+#### Applicability of this OCR bird-naming workflow
+
+Actually, the Biodiversity Heritage Library itself provides OCR output
+for its collection, see [this
+example](https://www.biodiversitylibrary.org/item/250938#page/37/mode/1up).
+I wasn’t able to find information about the software powering this OCR.
+What I *was* able to find out is that the BHL uses [purposeful
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: define abbrev BHL above and use consistently below if you use it >= 3x

File: content/blog/2018-08-28-birds-ocr.md [hunk 22]
```diff
@@ -0,0 +1,411 @@
...
+gaming](https://biodivlib.wikispaces.com/Purposeful+Gaming) in [its OCR
+workflow](https://biodivlib.wikispaces.com/file/view/Workflowdiagram.pdf).
+The raw OCR results aren’t much better than what we got in this post
+which is comforting.
+
+#### More data from the Biodiversity Heritage Library
+
+If you’re interested in the data, images but not only, of the
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: suggested rephrase:

If you’re interested in other types of data from the BHL, in addition to the images,

File: content/blog/2018-08-28-birds-ocr.md [hunk 23]
```diff
@@ -0,0 +1,411 @@
...
+```
+
+    ## [1] "Handbook to the birds of Australia. : [Supplementary material in Charles Darwin's copy]."
+    ## [2] "An introduction to The birds of Australia /"                                             
+    ## [3] "The useful birds of southern Australia : with notes on other birds /"                    
+    ## [4] "The Birds of Australia"                                                                  
+    ## [5] "The birds of Australia,"                                                                 
+    ## [6] "The birds of Australia,"
+
+And to get the OCR results of the pages of the book we used, we could
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: can you rephrase toclarify whether you are looking for page numbers?

File: content/blog/2018-08-28-birds-ocr.md [hunk 24]
```diff
@@ -0,0 +1,411 @@
...
+    ## [3] "APHELOCEPHALA LEUCOPSIS \n( WHITE FACE ). \n"                                                                                                                                                                                       
+    ## [4] "* \nAPHELOCEPHALA PE CTORALIS. \n(CHE <3 TNUT -BREASTED WHITEFA CEj. \nAPHELOCEPHALA NIGRICINCTA. \n(BE A CK-BAH.DED WHITE FA CEj. \n"                                                                                              
+    ## [5] "H . Gronvold. del. \nWitherLy & C° \nSPHENOSTOMA CRIS TATUM \n(WEDGE BIEL). \n"                                                                                                                                                     
+    ## [6] "49 6 \nH \n(jronvolcl. del. \nN E O SIT TA LE CJ C O CE PHAI.A. \n( WHITE ¦ HE AID EE THE EH UN HE FL). \nNEOSHTA ALBATA \n(F IE E> T Ft E EE UNNEFlj. \nNEOSITTA CHRYSOPTERA \nf OFi. A. NGE - wing-e d tree runner). \nWitWLjA \n"
+
+So there’s quite a lot to explore!
+
```
Comment (stefaniebutland) [inline content/blog/2018-08-28-birds-ocr.md]: same line break problem

File: content/blog/2018-08-28-birds-ocr.md (hunk 5 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: I'm rephrasing a bit but not sure I made the choice in the right direction of changes.

File: content/blog/2018-08-28-birds-ocr.md (hunk 13 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: 👍

Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: Ah no I re-checked taxize's maintainer is @sckott, it's `taxa` that got a new maintainer.

File: content/blog/2018-08-28-birds-ocr.md (hunk 18 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: I can't because it's just a random idea I wrote. Will add "untested".

File: content/blog/2018-08-28-birds-ocr.md (hunk 1 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: rephrased

File: content/blog/2018-08-28-birds-ocr.md (hunk 2 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: done

File: content/blog/2018-08-28-birds-ocr.md (hunk 3 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: 👍

File: content/blog/2018-08-28-birds-ocr.md (hunk 4 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: good point

File: content/blog/2018-08-28-birds-ocr.md (hunk 7 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: rephrased

File: content/blog/2018-08-28-birds-ocr.md (hunk 8 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: whyyy are they so much trouble?! I didn't succeed in getting them right :sob:

Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: I tried `<br>` and before that I had tried comparing them to working headers, no luck.

File: content/blog/2018-08-28-birds-ocr.md (hunk 9 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: 👍

File: content/blog/2018-08-28-birds-ocr.md (hunk 10 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: ✔️

File: content/blog/2018-08-28-birds-ocr.md (hunk 11 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: ✔️

File: content/blog/2018-08-28-birds-ocr.md (hunk 12 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: I split the chunk and added more comments, is it enough?

File: content/blog/2018-08-28-birds-ocr.md (hunk 14 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: ✅

File: content/blog/2018-08-28-birds-ocr.md (hunk 15 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: re-phrased

File: content/blog/2018-08-28-birds-ocr.md (hunk 16 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: ✅

File: content/blog/2018-08-28-birds-ocr.md (hunk 17 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: ✅

File: content/blog/2018-08-28-birds-ocr.md (hunk 19 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: ✅

File: content/blog/2018-08-28-birds-ocr.md (hunk 20 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: ✅

File: content/blog/2018-08-28-birds-ocr.md (hunk 21 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: ✅  except in a header of the conclusion to make it easier to skim

File: content/blog/2018-08-28-birds-ocr.md (hunk 23 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: I added comments in the code chunk

**Comment by maelle** [comment]
Many thanks @stefaniebutland , super useful as usual.

I was unable to fix headers. @jeroen do you have any idea why the headers

* Text extraction
* "More birding soon!" at the very end

don't start on a new line?! I added `<br>`, and had tested other small things earlier, with no success.

**Comment by jeroen** [comment]
It looks like displaying `article h4` headers as inline has been a [style choice](https://github.com/ropensci/roweb2/blob/master/themes/ropensci/static/css/sections.pcss#L265) by the designer. Maybe we should just remove that?

**Comment by maelle** [comment]
Maybe I'm being lazy but I think it could be removed, why would one want inline headers?

**Comment by jeroen** [comment]
@maelle I agree, I'm just going to remove that.

**Comment by jeroen** [comment]
Looks good to me, cool showcase of combination of ropensci packages.

**Comment by maelle** [comment]
Thanks @jeroen! Then after my vacation I'll "just" need to rebase the commit 😹

File: content/blog/2018-08-28-birds-ocr.md (hunk 13 above)
Comment (maelle) [reply content/blog/2018-08-28-birds-ocr.md]: np, it was a good remark
